File: src/main/resources/templates/go/solution/mergesort.go
package ${packageName}

import "time"

type MergeSort struct{}

func NewMergeSort() *MergeSort {
	return new(MergeSort)
}

// PerformSort is a wrapper method for the real MergeSort algorithm.
func (m *MergeSort) PerformSort(input []time.Time) {
	mergeSort(input)
}

// mergeSort recursively applies the MergeSort algorithm.
func mergeSort(input []time.Time) {
	if len(input) < 2 {
		return
	}

	middle := len(input) / 2
	mergeSort(input[:middle])
	mergeSort(input[middle:])
	merge(input, middle)
}

// merge merges two ranges within input defined from low to mid and from mid+1 to high.
func merge(input []time.Time, mid int) {
	temp := make([]time.Time, len(input))

	left, right, k := 0, mid, 0

	for left < mid && right < len(input) {
		if input[left].Before(input[right]) {
			temp[k] = input[left]
			left++
		} else {
			temp[k] = input[right]
			right++
		}
		k++
	}

	for left < mid {
		temp[k] = input[left]
		left++
		k++
	}
	for right < len(input) {
		temp[k] = input[right]
		right++
		k++
	}

	copy(input, temp)
}

File: src/main/resources/templates/ruby/solution/src/merge_sort.rb
# frozen_string_literal: true

class MergeSort
  # Wrapper method for the real MergeSort algorithm.
  # @param input [Array<Date>] the Array of Dates to be sorted
  def perform_sort(input)
    mergesort(input, 0, input.length)
  end

  private

  # Recursive merge sort method
  def mergesort(input, low, high)
    return if high - low <= 1

    mid = (low + high) / 2

    mergesort(input, low, mid)
    mergesort(input, mid, high)
    merge(input, low, mid, high)
  end

  # Merge method
  def merge(input, low, middle, high)
    temp = Array.new(high - low)

    left_index = low
    right_index = middle
    whole_index = 0

    while left_index < middle && right_index < high
      if input[left_index] <= input[right_index]
        temp[whole_index] = input[left_index]
        left_index += 1
      else
        temp[whole_index] = input[right_index]
        right_index += 1
      end
      whole_index += 1
    end

    while left_index < middle
      temp[whole_index] = input[left_index]
      left_index += 1
      whole_index += 1
    end

    while right_index < high
      temp[whole_index] = input[right_index]
      right_index += 1
      whole_index += 1
    end

    temp.each_with_index do |value, index|
      input[low + index] = value
    end
  end
end
